### PR TITLE
Paper: Papyri: Better documentation for the Scientific Ecosystem in Jupyter

### PR DESCRIPTION
**Pull request by Carreau**
See http://procbuild.scipy.org/ for logs generated by the build process.

Thanks for all your work on organising SciPy.

*Editor*: @stargaser

*Reviewers*: @wd15, @karthikmurugadoss

### CODE REVIEW

**Comment by scoobies** [comment]
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

**Comment by deniederhut** [comment]
@scoobies mark paper

**Comment by deniederhut** [comment]
@scoobies build status

**Comment by scoobies** [comment]
{"Carreau-2022"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpza_d6ary'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: bussonnier\n*** Warning: /tmp/tmpza_d6ary/publisher/../output/bussonnier/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/Carreau-2022.pdf", "build_status"=>"success", "build_timestamp"=>"27/05 20:19"}, "status"=>"success"}}

**Comment by deniederhut** [comment]
@scoobies mark needs review

**Comment by deniederhut** [comment]
@scoobies assign @stargaser as editor

**Comment by scoobies** [comment]
Assigned! @stargaser is now the editor

**Comment by wd15** [comment]
Hi, I've been asked to review this. Looking forward to it. Deadline for my input in June 9.

**Comment by stargaser** [comment]
@scoobies add @wd15 to reviewers

**Comment by scoobies** [comment]
@wd15 added to the reviewers list!

**Comment by deniederhut** [comment]
@scoobies add @karthikmurugadoss to reviewers

**Comment by scoobies** [comment]
@karthikmurugadoss added to the reviewers list!

**Comment by deniederhut** [comment]
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

**Comment by wd15** [comment]
Papyri is addressing a real need for improved documentation procedure
in scientific open-source development. A model that separates the
generation of API / example documentation from the view of the
document would be a major step forward in standardizing documentation
procedure across IDEs, web views and applications. This will greatly
reduce the document maintenance overhead required when developing
open-source software.

Overall, the paper gives a good introduction into the thinking behind
Papyri. These are my initial thoughts on how to improve the paper.

 - The grammar seems a bit loose in places. I can go through as best I
   can and mark where I notice grammar issues, but I suggest the
   authors give this another go or maybe pass it through a grammar
   engine. I'll look through it again once you've given it another
   read through.

 - The intermediate format or IRD is a very important step for the
   community. Other tools can build from this format either by
   generating the documentation view or by generating the IRD from the
   source code. It would be nice in the paper if the authors could
   actually describe the details of the format or schema for the
   IRD. The schema itself could become a standard for documentation
   and, thus, making it transparent to the reader would be useful.

 - Suggestion: Pandoc is a tool that uses an AST and can covert
   between many markup and documentation formats. Would it be useful
   to mention Pandoc in the paper as an example of a successful tool
   that uses a similar approach?

 - Suggestion: As part of the IRD, package metadata is stored. Would
   it be useful to use an existing schema such as [CODEMETA.yaml](https://codemeta.github.io/) for
   this.

 - Section numbering could be improved by using "X.Y" for
   subsections. Also, shouldn't the Introduction be numbered as Section 1?
 
I'll give you a chance to clean up the paper and give a more detailed
line-by-line review over the next week.

**Comment by deniederhut** [comment]
@scoobies mark needs more review

**Comment by karthikmurugadoss** [comment]
The paper presented here discusses Papyri which is an approach for unifying the documentation experience in the scientific Python ecosystem. Primarily, Papyri aims at decoupling the process of creating documentation (using the presented IRD format) and actually rendering the documentation (which can often times be user specific)

The paper describes the motivation and current state of Papyri quite well. The following are my major and minor comments. 

Major Comments:
1. Section on current implementation conveys a lot of information and is a bit difficult to follow for someone who isn't well-versed with existing documentation workflows. A few concepts that are introduced are RST parsing, CBOR representation, etc. 
2. The section on `IRD file installation` can benefit from a visual showing the different components (SQLite database, Raw storage on disk, etc.) and how these components interact with each other. 
3. Related to the above point, it is not clear why the object information is stored in 3 different places and what exactly is stored in each of these locations. More clarity here would be very helpful.
4. The context and usefulness of the local graph visualization is not described. Does Papyri create this visual as a part of the documentation generation process? Or is it here specifically to highly the connections into/from `ndarray`? 

Minor Comments:
1. There are typos and missing words in a number of places which need to be addressed. 
2. In the Current Implementation section, there are a number of referenced tools for which links to their sources would be helpful. e.g. Jedi, Pygments, Quart, Trio, etc. 
3. The numbering for sections can be improved

**Comment by deniederhut** [comment]
@scoobies mark pending comment

**Comment by Carreau** [comment]
Many thanks for both reviews, I've tried to address most of the above points in different commits to try to make re-review easier. 

Here is a small summary of the changes.

> The intermediate format or IRD is a very important step for the
> community. Other tools can build from this format either by
> generating the documentation view or by generating the IRD from the
> source code. It would be nice in the paper if the authors could
> actually describe the details of the format or schema for the
> IRD. The schema itself could become a standard for documentation
> and, thus, making it transparent to the reader would be useful.


I believe the project is too young to give a complete description of the IRD,
there are still regular changes to the format every 2 to 3 weeks depending on
the activity, thus a detail description would be premature. The IRD is still
changing much less frequently than initially but still too frequently IMHO. 
I tied to clarify this.


> Suggestion: Pandoc is a tool that uses an AST and can covert
> between many markup and documentation formats. Would it be useful
> to mention Pandoc in the paper as an example of a successful tool
> that uses a similar approach?

Yes, this is a good tool, I've added it. I was also recently made aware of
https://markdoc.io/docs/nodes, that was released recently but haven't had a
change to try it, so didn't wanted to make major changes that late in the
proceeding process. 

> Suggestion: As part of the IRD, package metadata is stored. Would
> it be useful to use an existing schema such as CODEMETA.yaml for
> this.

I've extended a bit this section, my main take so far is to limit to the minimum
vital, and rely on metadata that is stored somewhere else. I would much prefer
something like codemeta or other JSON LD format to be part of the package on
PyPI. 

> Section numbering could be improved by using "X.Y" for
> subsections. Also, shouldn't the Introduction be numbered as Section 1?

We removed section numbering altogether, it should be something that is handled
at the proceeding level. Is the proceeding is compiled with a directive like
`sectnum` then sections will have number, if we add them manually then the
number will appear twice.



> Section on current implementation conveys a lot of information and is a bit
> difficult to follow for someone who isn't well-versed with existing
> documentation workflows. A few concepts that are introduced are RST parsing,
> CBOR representation, etc.


I've extend where possible, but as for other comments above about IRD schema, I
don't want to dive too far into the implementation as it is still changing
quickly and should not be necessary to understand the idea.

I hope that in the future other competing projects will for that
produce/consume IRD bundles, and potentially make completely different technical
choices.

> The section on IRD file installation can benefit from a visual showing the
> different components (SQLite database, Raw storage on disk, etc.) and how
> these components interact with each other.

I've tried to clarify as well, and made a schema. I've also extended that these
choices are mostly made due to the current use case I target and could/should be
reconsider by an implementation with different targets.

> Related to the above point, it is not clear why the object information is
> stored in 3 different places and what exactly is stored in each of these
> locations. More clarity here would be very helpful.

As for above, I've tried to clarify, let me know if this is clearer.

> The context and usefulness of the local graph visualization is not described.
> Does Papyri create this visual as a part of the documentation generation
> process? Or is it here specifically to highly the connections into/from
> ndarray?


Thanks, this was indeed not clear, I've reworked this section. This was trying
to  demonstrate that changes to the documentation could be done without having
to re-do the generation step. This graph is indeed generated at render time, and
updates depending on which libraries you have documentation installed for, and
give you ideas of types of UI changes that could be implemented later. 


> There are typos and missing words in a number of places which need to be addressed.

We tried to fix things the best we could, and would appreciate any pointers to
remaining mistakes.

> In the Current Implementation section, there are a number of referenced tools for which links to their sources would be helpful. e.g. Jedi, Pygments, Quart, Trio, etc.


This should be mostly fixed, beyond a couple of citation I need to expand with
the right DOIs.

**Comment by deniederhut** [comment]
Awesome! @wd15 and @karthikmurugadoss -- do you feel this paper is now ready for inclusion in the proceedings?

**Comment by karthikmurugadoss** [comment]
Yes did another read through and it looks good to me!

**Comment by Carreau** [comment]
> Yes did another read through and it looks good to me!

Many thanks, that's giving me extra motivation after the week-end !

**Comment by wd15** [comment]
Looks much better. Nice work!

**Review by wd15 — APPROVED** [review]

**Review by karthikmurugadoss — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -477,6 +477,7 @@ Editors should come to a final 'ready', 'unready' decision before the **Final Ed
 5. Choose the appropriate folder, and make a new directory inside it (it needs a unique name)
 6. Copy your slide deck or poster into the directory, and add a file called `info.json` with the following fields needed for uploading to Zenodo (using an empty string for author orcid or
 affiliation if these cannot be provided):
+
```
Comment (stargaser) [inline README.md]: @Carreau would you please remove the edit to the top-level README.md file? Files outside the papers directory cannot be modified.

Comment (Carreau) [inline README.md]:
```suggestion
```

Comment (Carreau) [reply README.md]: Oops, sorry about that. Should be removed now. Should I rebase to make sure it is not even touched in the history ?

Comment (stargaser) [reply README.md]: Thanks @Carreau, I think there is no need to rebase.

**Comment by stargaser** [comment]
@scoobies mark ready

**Pull request merged**